File: mergsort.py
def merge_sort(array, start, end):
    aux_array = array.copy()
    if (len(array) == 0 or len(array) == 1):
        return
    split(array, start, end, aux_array) 

def split(array, start, end, aux_array):
    middle = (start + end) // 2
#    print(start, middle, end)
    if (start == end):
        return
    split(array, start, middle, aux_array) 
    split(array, middle + 1, end, aux_array)
    merge(array, start, middle, end, aux_array)

def merge(array, start, middle,  end, aux_array):
    i = start
    k = start
    j = middle + 1
    
    while(i <= middle and j <= end):
        if (aux_array[i] < aux_array[j]):
            array[k] = aux_array[i]
            k+=1
            i+=1
        else:
            array[k] = aux_array[j]
            k+=1
            j+=1

    while (i <= middle):
        array[k] = aux_array[i]
        k+=1
        i+=1

    while (j <= end):
        array[k] = aux_array[j]
        k+=1
        j+=1


if __name__ == '__main__':
    array = [3, 2, 1]
    merge_sort(array, 0, len(array)-1)
    print(array)

File: mergsort.ts
export function MergSort(
  array: number[],
) {
  if (array.length <= 1) return array
  const auxArray = array.slice()
  Split(array, 0, array.length - 1, auxArray)
}

function Split(
  array: number[],
  start: number,
  end: number,
  auxArray: number[],
) {
  if (start === end) return
  const middle = Math.floor((start + end) / 2)
  Split(auxArray, start, middle, array)
  Split(auxArray, middle + 1, end, array)
  merge(array, start, middle, end, auxArray)
}

function merge(
  array: number[],
  start: number,
  middle: number,
  end: number,
  auxArray: number[],
) {
  let k = start
  let i = start
  let j = middle + 1
  while (i <= middle && j <= end) {
    if (auxArray[i] <= auxArray[j]) {
      array[k++] = auxArray[i++]
    } else {
      array[k++] = auxArray[j++]
    }
  }
  while (i <= middle) {
    array[k++] = auxArray[i++]
  }
  while (j <= end) {
    array[k++] = auxArray[j++]
  }
}

let array = [5, 4, 3, 2, 1]
MergSort(array)
console.log(array)
